native/acter/src/api/verification.rs

```diff
@@ -1105,11 +1110,217 @@ impl VerificationController {
+                    .encryption()
+                    .get_user_devices(user_id)
+                    .await
+                    .context("Couldn't get crypto devices")?;
+                let mut sessions = vec![];
+                for device in response.devices {
+                    let is_verified = crypto_devices.get(&device.device_id).is_some_and(|d| {
+                        d.is_cross_signed_by_owner() || d.is_verified_with_cross_signing()
+                    });
+                    if !is_verified {
+                        sessions.push(DeviceRecord::new(
+                            device.device_id.clone(),
+                            device.display_name.clone(),
+                            device.last_seen_ts,
+                            device.last_seen_ip.clone(),
+                            false,
+                        ));
+                    }
+                }
+                warn!("unverified sessions: {:?}", sessions);
+                Ok(sessions)
+            })
+            .await?
+    }
+
+    pub async fn inactive_sessions(&self) -> Result<Vec<DeviceRecord>> {
+        let client = self.client.clone();
+        RUNTIME
+            .spawn(async move {
+                let user_id = client.user_id().context("User not found")?;
+                let response = client.devices().await?;
+                let crypto_devices = client
+                    .encryption()
+                    .get_user_devices(user_id)
+                    .await
+                    .context("Couldn't get crypto devices")?;
+                let mut sessions = vec![];
+                for device in response.devices {
+                    let mut is_inactive = true;
+                    if let Some(last_seen_ts) = device.last_seen_ts {
+                        let limit = SystemTime::now()
+                            .checked_sub(Duration::from_secs(90 * 24 * 60 * 60))
+                            .context("Couldn't get time of 90 days ago")?
+                            .duration_since(UNIX_EPOCH)
+                            .context("Couldn't calculate duration from Unix epoch")?;
+                        let secs: u64 = last_seen_ts.as_secs().into();
+                        if secs > limit.as_secs() {
+                            is_inactive = false;
+                        }
+                    }
+                    if is_inactive {
+                        let is_verified = crypto_devices.get(&device.device_id).is_some_and(|d| {
+                            d.is_cross_signed_by_owner() || d.is_verified_with_cross_signing()
+                        });
+                        sessions.push(DeviceRecord::new(
+                            device.device_id.clone(),
+                            device.display_name.clone(),
+                            device.last_seen_ts,
+                            device.last_seen_ip.clone(),
+                            is_verified,
+                        ));
```

Comment:
these three methods are the same aside from a single check each. How about we also store the `is_inactive` on the `DeviceRecord` and calculate that as well. Then we can just have one `.sessions` method and implement it and the other two remaining methods `delete_devices` and `request_session` directly on `Client` rather than on this unnecessary in between stud ...

```suggestion
impl Client {
    pub async fn sessions(&self) -> Result<Vec<DeviceRecord>> {
        let client = self.client.clone();
        RUNTIME
            .spawn(async move {
                let user_id = client.user_id().context("User not found")?;
                let response = client.devices().await?;
                let crypto_devices = client
                    .encryption()
                    .get_user_devices(user_id)
                    .await
                    .context("Couldn't get crypto devices")?;
                let mut sessions = vec![];
                for device in response.devices {
                    let mut is_active = false;
                    if let Some(last_seen_ts) = device.last_seen_ts {
                        let limit = SystemTime::now()
                            .checked_sub(Duration::from_secs(90 * 24 * 60 * 60))
                            .context("Couldn't get time of 90 days ago")?
                            .duration_since(UNIX_EPOCH)
                            .context("Couldn't calculate duration from Unix epoch")?;
                        let secs: u64 = last_seen_ts.as_secs().into();
                        if secs > limit.as_secs() {
                            is_active = true;
                        }
                    }
                      let is_verified = crypto_devices.get(&device.device_id).is_some_and(|d| {
                          d.is_cross_signed_by_owner() || d.is_verified_with_cross_signing()
                      });
                      sessions.push(DeviceRecord::new(
                          device.device_id.clone(),
                          device.display_name.clone(),
                          device.last_seen_ts,
                          device.last_seen_ip.clone(),
                          is_verified,
                          is_active,
                      ));
```

Reply:
I wanted to avoid making `Client` crate too big.
So I implemented `all_sessions()`, `verified_sessions()`, `unverified_sessions()`, `inactive_sessions()`, `delete_devices()` and `request_verification()` in `SessionManager` not `Client`.

Reply:
though my patch shows that we don't need that many functions and the number of functions on the client struct doesn't really matter - as we can split it up into various modules (like we do here), it makes it feasible to handle. While having a secondary struct that needs to hold another clone-able client internally is just more cumbersome to use and uses more resources (extra allocation for the additional clone). Arguably this isn't much, but unless we have actual internal state (like the device-record has), I prefer keeping the layers low rather than nested.

Reply:
[This issue](https://github.com/acterglobal/a3/issues/789) will not be closed yet.
Because I have to implement `KeyVerificationReady` event handling in cross signing.
At that time I can delete `SessionManager` crate, if you want to remove `SessionManager` crate and let `Client` replace `SessionManager`.